Add MINILMV2

LiuChiachi opened the pull request:
### PR types
<!-- One of [ New features | Bug fixes | Function optimization | Performance optimization | Breaking changes | Others ] -->
New features
### PR changes
<!-- One of [ Models | APIs | Docs | Others ] -->
Models
### Description
<!-- Describe what this PR does -->
A implementation of MINILMv2, a general distillation algorithm, and update `distill_utils.py` to support MINILM.
And it supports distributed training.

Referenced this [issue](https://github.com/microsoft/unilm/issues/354)

LiuChiachi committed:
add minilm task distill

LiuChiachi committed:
Merge branch 'develop' of https://github.com/PaddlePaddle/PaddleNLP into add-minilm

LiuChiachi committed:
add general distillation MINILMV2, and update distill_utils

LiuChiachi committed:
Merge branch 'develop' of https://github.com/PaddlePaddle/PaddleNLP into add-minilm

ZeyuChen reviewed (commented):
要注意同步更新模型文档列表

LiuChiachi committed:
remove test break and gradient accumulate

LiuChiachi committed:
fix loss funct

LiuChiachi committed:
optimize distill utils

LiuChiachi committed:
Merge branch 'develop' of https://github.com/PaddlePaddle/PaddleNLP into add-minilm

LiuChiachi committed:
Merge branch 'develop' of https://github.com/PaddlePaddle/PaddleNLP into add-minilm

LiuChiachi committed:
add tinybert general distillation

LiuChiachi committed:
Merge branch 'develop' of https://github.com/PaddlePaddle/PaddleNLP into add-minilm

LiuChiachi committed:
add tinybert general distillation

LiuChiachi committed:
update default hyperparameters

LiuChiachi committed:
Merge branch 'develop' of https://github.com/PaddlePaddle/PaddleNLP into add-minilm

LiuChiachi committed:
add cold boot

LiuChiachi committed:
Merge branch 'develop' of https://github.com/PaddlePaddle/PaddleNLP into add-minilm

LiuChiachi committed:
Merge branch 'add-minilm' of https://github.com/LiuChiachi/PaddleNLP into add-minilm

tianxin1860 commented on examples/model_compression/minilm/general_distill.py:
```diff
@@ -0,0 +1,413 @@
+    warmup = args.warmup_steps if args.warmup_steps > 0 else args.warmup_proportion
+    lr_scheduler = LinearDecayWithWarmup(args.learning_rate, num_training_steps,
+                                         warmup)
+    optimizer = paddle.optimizer.Adam(
```
Adam with WeightDecay 的实现不应该使用这个 API  吧,  替换成 AdamW ?

tianxin1860 replied:
Adam 和 weight_decay 结合起来使用的时候，并不是所有参数都需要 weight_decay: bias、norm 相关参数不需要 decay.

tianxin1860 replied:
AdamW optimizer 在使用的时候最好加入梯度裁剪 Clip 限制梯度最大 Norm， 否则训练过程中遇见噪音数据可能模型就发散了.

LiuChiachi replied:
Done. Thanks!

tianxin1860 reviewed (commented):
Optimizer 部分需要一些修改

LiuChiachi committed:
correct optimizer adam to adamw using weight decay, and add clip norm

LiuChiachi committed:
Merge branch 'develop' of https://github.com/PaddlePaddle/PaddleNLP into add-minilm

LiuChiachi committed:
Merge branch 'add-minilm' of https://github.com/LiuChiachi/PaddleNLP into add-minilm

LiuChiachi committed:
update optimizer for minilm, and add clip norm

LiuChiachi committed:
Merge branch 'develop' of https://github.com/PaddlePaddle/PaddleNLP into add-minilm

tianxin1860 commented on examples/model_compression/minilm/general_distill.py:
```diff
@@ -0,0 +1,423 @@
+    lr_scheduler = LinearDecayWithWarmup(args.learning_rate, num_training_steps,
+                                         warmup)
+
+    clip = paddle.nn.ClipGradByGlobalNorm(clip_norm=1.0)
```
尽量避免硬编码，`clip_norm = args.max_grad_norm` 合适一些.

LiuChiachi replied:
Done. Thanks!

tianxin1860 commented on examples/model_compression/minilm/general_distill.py:
```diff
@@ -0,0 +1,423 @@
+    pool = ThreadPoolExecutor(1)
+
+    teacher = to_distill(teacher, return_qkv=True, layer_index=18)
+    student = to_distill(student, return_qkv=True, layer_index=5)
```
layer_index 参数是否考虑拆分作为 teacher_layer_index 和 student_ layer_index 从命令行传入比较合适？

LiuChiachi replied:
Done. Thanks!

tianxin1860 commented on examples/model_compression/minilm/general_distill.py:
```diff
@@ -0,0 +1,423 @@
+            remainder = paddle.distributed.get_world_size() % num_files
+
+            data_file = files[(
+                f_start_id * paddle.distributed.get_world_size() +
```
f_start_id 的取值始终为 0 ？

tianxin1860 reviewed (changes requested):
Leave some comment

tianxin1860 commented on examples/model_compression/minilm/general_distill.py:
```diff
@@ -0,0 +1,423 @@
+
+    # Required parameters
+    parser.add_argument(
+        "--model_type",
```
是否命名为 student_model_type 与 teacher_model_type 相对应?

LiuChiachi replied:
Done. Thanks!

tianxin1860 commented on examples/model_compression/tinybert/general_distill.py:
```diff
@@ -0,0 +1,451 @@
+        random.seed(self.seed + id)
+
+
+def create_pretraining_dataset(input_file, shared_list, args, worker_init,
```
`shared_list` is an unused variable

LiuChiachi replied:
Done. Thanks!

tianxin1860 commented on examples/model_compression/minilm/general_distill.py:
```diff
@@ -0,0 +1,423 @@
+        random.seed(self.seed + id)
+
+
+def create_pretraining_dataset(input_file, shared_list, args, worker_init,
```
`shared_list` is an unused variable

LiuChiachi replied:
Done. Thanks!

tianxin1860 commented on examples/model_compression/minilm/general_distill.py:
```diff
@@ -0,0 +1,423 @@
+                                         shared_file_list, args, worker_init,
+                                         tokenizer)
+
+            kl_loss_fct = paddle.nn.KLDivLoss('sum')
```
`kl_loss_fct` -> `kl_loss_func` 表意更清楚一些.

LiuChiachi replied:
Done. Thanks!

tianxin1860 commented on examples/model_compression/tinybert/general_distill.py:
```diff
@@ -0,0 +1,451 @@
+        apply_decay_param_fun=lambda x: x in decay_params,
+        grad_clip=clip)
+
+    ce_loss_fct = paddle.nn.CrossEntropyLoss(soft_label=True)
```
`kl_loss_fct` -> `kl_loss_func`

LiuChiachi replied:
Done. Thanks!

tianxin1860 commented on examples/model_compression/tinybert/general_distill.py:
```diff
@@ -0,0 +1,451 @@
+        grad_clip=clip)
+
+    ce_loss_fct = paddle.nn.CrossEntropyLoss(soft_label=True)
+    mse_loss_fct = paddle.nn.MSELoss()
```
同上

tianxin1860 commented on examples/model_compression/tinybert/general_distill.py:
```diff
@@ -0,0 +1,451 @@
+            for step, batch in enumerate(train_data_loader):
+                global_step += 1
+                input_ids = batch[0]
+                attention_mask = paddle.unsqueeze(
```
`attention_mask ` is unused ？

tianxin1860 replied:
1. 此处的 attention_mask 计算出来之后为什么不需要传到前向 forward 函数里？
2. 如果不传的话，此处没必要计算 attention_mask.

tianxin1860 commented on examples/model_compression/tinybert/task_distill.py:
```diff
@@ -378,16 +383,17 @@ def do_train(args):
+    teacher = to_distill(
+        teacher,
+        return_attentions=True,
+        return_qkv=True,
```
应该是 `return_qkv=False` ?

LiuChiachi replied:
Done. Thanks!

tianxin1860 commented on examples/model_compression/tinybert/task_distill.py:
```diff
@@ -378,16 +383,17 @@ def do_train(args):
+    student = to_distill(
+        student,
+        return_attentions=True,
+        return_qkv=True,
```
同上

LiuChiachi replied:
Done. Thanks!

tianxin1860 commented on examples/model_compression/tinybert/task_distill.py:
```diff
@@ -416,7 +426,10 @@ def cal_intermediate_distill_loss(student, teacher):
             global_step += 1
             input_ids, segment_ids, labels = batch
-
+            attention_mask = paddle.unsqueeze(
```
计算了 attention_mask 没有使用

LiuChiachi replied:
Deleted. Thanks!

tianxin1860 commented on paddlenlp/transformers/distill_utils.py:
```diff
@@ -15,35 +14,97 @@
 
 
-def to_distill(self):
+def calc_minilm_loss(loss_fct, s, t, attn_mask, num_relation_heads=0):
```
calc_minilm_loss 的逻辑属于 MiniLM 策略的一部分，是否放在 MiniLM 目录下实现合适一些？ distill_utils.py 我理解应该是一些模型无关的通用逻辑实现?

LiuChiachi replied:
希望不止是在这个example里可以用到minilmv2策略

tianxin1860 commented on paddlenlp/transformers/distill_utils.py:
```diff
@@ -146,20 +216,83 @@ def transformer_encoder_forward(self, src, src_mask=None, cache=None):
     self.hidden_states = []
-    self.v = None
+
     for i, mod in enumerate(self.layers):
```
`mod` 变量命名表意不清楚

LiuChiachi replied:
这里参考了框架transformer API的写法。

tianxin1860 reviewed (changes requested):
Leave some comment

LiuChiachi committed:
update arg and support minilmv2

LiuChiachi committed:
Merge branch 'develop' of https://github.com/PaddlePaddle/PaddleNLP into add-minilm

LiuChiachi committed:
add readme for minilmv2

add run_clue.py scripts

update dataset name of wsc

update dataset name of wsc

add parameters link

tianxin1860 commented on examples/model_compression/minilm/README.md:
```diff
@@ -0,0 +1,68 @@
+
+### 训练启动方式
+
+假设我们把数据文件都放在`/PaddleNLP/dataset`下，那么我们可以运行如下命令用单机八卡进行预训练蒸馏：
```
CLUE 数据集不是已经集成到 PaddleNLP dataset 里了吗？

tianxin1860 reviewed (commented):
Leave a comment

tianxin1860 reviewed (approved):
LGTM

LiuChiachi committed:
Merge branch 'develop' into add-minilm
